src/scope.rs

```diff
@@ -0,0 +1,232 @@
+    {
+        let scope = &Scope { namespace: self, bag: bag };
+        f(scope)
+    }
+
+    #[inline]
+    unsafe fn unprotected<F, R>(self, f: F) -> R where
+        F: FnOnce(&Scope<Self>) -> R,
+    {
+        let mut bag = Bag::new();
+        let result = self.unprotected_with_bag(&mut bag, f);
+        drop(bag); // FIXME(jeehoonkang)
+        result
+    }
+}
+
+
+pub struct Agent<'scope, N: Namespace + 'scope> {
+    /// This agent's namespace
+    namespace: N,
+    /// This agent's entry in the registry list.
+    registry: &'scope Node<Registry>,
+    /// The local garbage objects that will be later freed.
+    bag: UnsafeCell<Bag>,
+    /// Whether the thread is currently pinned.
+    is_pinned: Cell<bool>,
+    /// Total number of pinnings performed.
+    pin_count: Cell<usize>,
+}
+
+impl<'scope, N> Agent<'scope, N> where
+    N: Namespace + 'scope,
+{
+    pub fn new(n: N) -> Self {
+        Agent {
+            namespace: n,
+            registry: unsafe {
+                // Since we don't dereference any pointers in this block, it's okay to use
+                // `unprotected`.  Also, we use an invalid bag since no garbages are created in list
+                // insertion.
+                let mut bag = ::std::mem::zeroed::<Bag>();
+                n.unprotected_with_bag(&mut bag, |scope| {
+                    &*n.registries().insert_head(Registry::new(), scope).as_raw()
+                })
+            },
+            bag: UnsafeCell::new(Bag::new()),
+            is_pinned: Cell::new(false),
+            pin_count: Cell::new(0),
+        }
+    }
+
+    /// Pins the current thread.
+    ///
+    /// The provided function takes a reference to a `Scope`, which can be used to interact with
+    /// [`Atomic`]s. The scope serves as a proof that whatever data you load from an [`Atomic`] will
+    /// not be concurrently deleted by another thread while the scope is alive.
+    ///
+    /// Note that keeping a thread pinned for a long time prevents memory reclamation of any newly
+    /// deleted objects protected by [`Atomic`]s. The provided function should be very quick -
+    /// generally speaking, it shouldn't take more than 100 ms.
+    ///
+    /// Pinning is reentrant. There is no harm in pinning a thread while it's already pinned (repinning
+    /// is essentially a noop).
+    ///
+    /// Pinning itself comes with a price: it begins with a `SeqCst` fence and performs a few other
+    /// atomic operations. However, this mechanism is designed to be as performant as possible, so it
+    /// can be used pretty liberally. On a modern machine pinning takes 10 to 15 nanoseconds.
+    ///
+    /// [`Atomic`]: struct.Atomic.html
+    pub fn pin<F, R>(&self, f: F) -> R where
+        F: FnOnce(&Scope<N>) -> R,
+    {
+        let registry = self.registry.get();
+        let scope = &Scope { namespace: self.namespace, bag: self.bag.get() };
+
+        let was_pinned = self.is_pinned.get();
+        if !was_pinned {
```

Comment:
Why not merge is_pinned and pin count, since it looks like is_pinned being true is equivalent to pin_count==0? Unless I'm misreading where the variables are used.